src/ray/rpc/gcs_server/gcs_rpc_client.h

```diff
@@ -596,138 +549,33 @@ class GcsRpcClient {
-      executor->Abort(
-          ray::Status::TimedOut("Timed out while waiting for GCS to become available."));
-      pending_requests_bytes_ -= request_bytes;
-      delete executor;
-      pending_requests_.erase(iter);
-    }
-
-    switch (status) {
-    case GRPC_CHANNEL_TRANSIENT_FAILURE:
-    case GRPC_CHANNEL_CONNECTING:
-      if (!gcs_is_down_) {
-        gcs_is_down_ = true;
-      } else {
-        if (absl::ToInt64Seconds(absl::Now() - gcs_last_alive_time_) >=
-            ::RayConfig::instance().gcs_rpc_server_reconnect_timeout_s()) {
-          RAY_LOG(ERROR) << "Failed to connect to GCS within "
-                         << ::RayConfig::instance().gcs_rpc_server_reconnect_timeout_s()
-                         << " seconds. "
-                         << "GCS may have been killed. It's either GCS is terminated by "
-                            "`ray stop` or "
-                         << "is killed unexpectedly. If it is killed unexpectedly, "
-                         << "see the log file gcs_server.out. "
-                         << "https://docs.ray.io/en/master/ray-observability/user-guides/"
-                            "configure-logging.html#logging-directory-structure. "
-                         << "The program will terminate.";
-          std::_Exit(EXIT_FAILURE);
-        }
-      }
-      break;
-    case GRPC_CHANNEL_SHUTDOWN:
-      RAY_CHECK(shutdown_) << "Channel shoud never go to this status.";
-      break;
-    case GRPC_CHANNEL_READY:
-    case GRPC_CHANNEL_IDLE:
-      gcs_last_alive_time_ = absl::Now();
-      gcs_is_down_ = false;
-      // Retry the one queued.
-      while (!pending_requests_.empty()) {
-        pending_requests_.begin()->second.first->Retry();
-        pending_requests_.erase(pending_requests_.begin());
-      }
-      pending_requests_bytes_ = 0;
-      break;
-    default:
-      RAY_LOG(FATAL) << "Not covered status: " << status;
-    }
-    SetupCheckTimer();
-  }
-
   const std::string gcs_address_;
   const int64_t gcs_port_;
-
-  instrumented_io_context *const io_context_;
-
-  // Timer can be called from either the GCS RPC event loop, or the application's
-  // main thread. It needs to be protected by a mutex.
-  absl::Mutex timer_mu_;
-  const std::unique_ptr<boost::asio::deadline_timer> timer_;
+  std::shared_ptr<grpc::Channel> channel_;
+  std::shared_ptr<RetryableGrpcClient> retryable_grpc_client_;
 
   /// The gRPC-generated stub.
-  std::unique_ptr<GrpcClient<JobInfoGcsService>> job_info_grpc_client_;
-  std::unique_ptr<GrpcClient<ActorInfoGcsService>> actor_info_grpc_client_;
-  std::unique_ptr<GrpcClient<NodeInfoGcsService>> node_info_grpc_client_;
-  std::unique_ptr<GrpcClient<NodeResourceInfoGcsService>> node_resource_info_grpc_client_;
-  std::unique_ptr<GrpcClient<WorkerInfoGcsService>> worker_info_grpc_client_;
-  std::unique_ptr<GrpcClient<PlacementGroupInfoGcsService>>
+  std::shared_ptr<GrpcClient<JobInfoGcsService>> job_info_grpc_client_;
```

Comment:
Why do we use `shared_ptr` here? IMO apart from rare cases (i.e. multi-threaded env, async) shared pointer should never been used.

Reply:
The only possibility of using shared pointer is when programmer is not sure about the lifecycle for certain object.

Reply:
Because retryable_grpc_client needs to take a shared pointer of grpc_client.